classes/local/step/connector_s3.php

```diff
@@ -29,12 +29,33 @@
  */
 class connector_s3 extends connector_step {
 
-    /** @var bool whether or not this step type (potentially) contains a side effect or not */
-    protected $hassideeffect = true;
-
     /** @var string the prefix identifier for an s3 path, e.g. s3://path/to/file. */
     const S3_PREFIX = 's3://';
 
+    /**
+     * Returns whether or not the step configured, has a side effect
+     *
+     * A side effect if it modifies some state variable value(s) outside its
+     * local environment, which is to say if it has any observable effect other
+     * than its primary effect of returning a value to the invoker of the
+     * operation.
+     *
+     * For s3 connectors, it is considered to have a side effect if the target is
+     * anywhere outside of the scratch directory.
+     *
+     * @return     bool whether or not this step has a side effect
+     * @link https://en.wikipedia.org/wiki/Side_effect_(computer_science)
```

Comment:
Not sure about any particular conventions, it can probably be discussed but I feel like the extra information is just unnecessary. The purpose of the original phpdocs was so step authors could read and know what the has_side_effect function is used for.

My edit would be as follows, which would keep the main title, and @return line. Everything else is tailored for this method's implementation.
```suggestion
     * For s3 connectors, it is considered to have a side effect if the target is
     * anywhere outside of the scratch directory.
     *
     * @return     bool whether or not this step has a side effect
```

Reply:
done